### PR TITLE
WIP:  React Week 1 - Hooks

### PR DESCRIPTION
**Pull request by CheezItMan**
## Description

I'm working through changing the React curriculum for week 1 into using Hooks instead of class components.  I've also tried to smooth things out a bit sticking with Ada Students throughout.  I've also written the [AdaGold version of AdaStudents](https://github.com/adagold/ada-students).  The master branch just has the create-react-app starter code, and there's a branch for each lecture in the process.  Each branch has the ending code for each lesson.

## Things left to do:

1.  Update the projects both solutions and starter code to call the `useState` hook.
    - Week 1 projects are done
1.  Look for additional resources on Hooks
1.  Update the schoology assessments

## Questions

1.  Should we modify Litter patrol, I just remember it being clunky

## Questions for the Author
- [ ] Have you updated the weekly learning goals in the pedagogy resource?
- [ x ] Have you updated the weekly assessment in Schoology?
- [ x ] Have you added links to the source (lucidchart, draw.io) for any diagram?
- [ x ] Have you updated the READMEs with links to new or moved resources?

## Feedback Requested
Please utilize the built-in GitHub Reviewers functionality to assign necessary reviewers. Use this section to further specify portions of the PR that you would like additional or focused feedback on.

### CODE REVIEW

File: React/class-components.md [hunk 1]
```diff
@@ -21,9 +21,19 @@ const Student = (props) => {
...
+
+### So Why Didn't We Learn Class Components Earlier?  
+
+At Ada we feel that, in most cases, that the recent introduction of hooks make class components _unnecessary_.  With hooks most applications can be written exclusively with functional components.  There are a few [specialized cases](https://reactjs.org/docs/error-boundaries.html) which require class components, but these scenarios are few and far between and may be replaced hook-based solutions at a later date.
+
+### So Why Learn About Class Components?
+
+React has been around for a while now and a great deal of code and documentation has been written using class components.  React is not eliminating class components and a great number of developers find the abstraction of class components attractive.  It's important to know how they work and how to convert between functional and class components.
```
Comment (jmaddox19) [inline React/class-components.md]: This is all really helpful context and good reasons to take the time to learn class components. 
That said, I wonder if it would be worth pulling this out of the curriculum, at least for C13 because:
- We are expecting to move at a little slower pace (from my understanding), so considering we are adding content about the useState hook I expect we'll need to cut something.
- I would argue that since our goal is to prepare them to be full-stack devs, not so much to be react devs specifically, we could expect that any students who need to can learn about class components on the job and many students will never need to.

File: React/creating-components.md [hunk 2]
```diff
@@ -103,6 +103,28 @@ We used the existing index.js code to render the `App` component that was provid
 
 In the next phase of this application, we are going to extend our application to support multiple students and switch from hard-coded to dynamic data.
 
+### An example Student Component
```
Comment (jmaddox19) [inline React/creating-components.md]: Thanks for doing all this work to create a more uniform flow with the student component!

File: React/state.md [hunk 3]
```diff
@@ -1,93 +1,186 @@
 # Managing data using `state`
 
+# TODO - Update for `useState` Hook
```
Comment (jmaddox19) [inline React/state.md]: Is this TODO done now or is there still more to do?

File: React/state.md [hunk 4]
```diff
@@ -1,93 +1,186 @@
...
-State is managed **within a given component**. State can be accessed using the `this.state` object and it can be modified using the `this.setState()` function.
 
-#### Overview
-- Can be assigned to any object key you want
-- `setState` will always take an object as the parameter
-- `setState` will merge any params passed in with the existing state data, it won't overwrite it
-- We often set the initial state in the constructor function
+State is managed **within a given component**. State can be accessed using the `useState` hook and it can be modified using the a function that `useState` returns.
+
+**Wait what's a hook?**  A hook is a special function react provides to "hook" into the lifecyle of a React component and access specific functionality.  The `useState` hook function allows us to have a variable, like an instance variable for a Ruby class for our component.  It also provides a function to change the value of that variable, or it's state.  When this state variable is changed, the component is re-rendered.  React has a number of hooks, and you can even create your own, but for now, we will focus on `useState`.
```
Comment (jmaddox19) [inline React/state.md]:
```suggestion
**Wait what's a hook?**  A hook is a special function react provides to "hook" into the lifecyle of a React component and access specific functionality.  The `useState` hook function allows us to have a variable with scope beyond the execution of a function. In Ruby, an instance variable served this same purpose.  It also provides a function to change the value of that variable (i.e. it's state).  When this state variable is changed, the component is re-rendered.  React has a number of hooks, and you can even create your own, but for now, we will focus on `useState`.
```

Comment (jmaddox19) [reply React/state.md]: This is my attempt to expand on the comparison to instance variables a little. But feel free to ignore if you don't feel it's more clear.

File: React/state.md [hunk 5]
```diff
@@ -1,93 +1,186 @@
...
+
+`useState` allows us to have components with data they remember internally and change over time, like instance variables in a Ruby object.
```
Comment (jmaddox19) [inline React/state.md]: Oh, this helps a whole lot to provide succinct clarity anyway.

File: React/state.md [hunk 6]
```diff
@@ -1,93 +1,186 @@
...
+#### useState Overview
+
+- will take the initial value of the state object as a parameter
+- will return an array with two elements
+  - the current value of the state
+  - a function to change the state
+
+## Adding State to the Student Component
+
+Lets say we want each student component to keep track of if the student is present or absent and we want to be able to change that value if they are present and re-render the component.
```
Comment (jmaddox19) [inline React/state.md]:
```suggestion
Lets say we want each student component to keep track of whether the student is present or absent and we want to be able to change that value if they are present and re-render the component.
```

File: React/state.md [hunk 7]
```diff
@@ -1,93 +1,186 @@
...
+### Calling useState
+
+We can then call `useState` in our Student component.
+
+```javascript
+// src/components/Student.js
+// ...
+const Student = (props) => {
+
+  const [present, setPresent] = useState(false);
```
Comment (jmaddox19) [inline React/state.md]: Would isPresent be more clear here? Maybe that's just the convention I'm used to in Java.
```suggestion
  const [isPresent, setIsPresent] = useState(false);
```

File: React/state.md [hunk 8]
```diff
@@ -1,93 +1,186 @@
...
+  // ...
+```
+
+When we called `useState` above we passed in the initial value of the state.  In this example we are defaulting students to being not present (absent).
+
+`useState` returns an array.  We could have written the above as `const presentArray = useState(false)` and then `presentArray[0]` would the the value and `presentArray[1]` would be a function to change the state, but again we are using a feature called destructuring to break that array into two variables `present` and `setPresent`.  This is a common technique when using hooks in React.
```
Comment (jmaddox19) [inline React/state.md]:
```suggestion
`useState` returns an array.  We could have written the above as `const presentArray = useState(false)` and then `presentArray[0]` would the the value and `presentArray[1]` would be a function we can use to change the state.
Instead, we are using a feature called destructuring (like we did with the import above) to break that array into two variables `present` and `setPresent`.  This is a common technique when using hooks in React.
```

File: React/state.md [hunk 9]
```diff
@@ -1,93 +1,186 @@
...
 ```
 
-Once the initial state is set, we can then make changes using the `setState` function. (**Note:** `setState` will work even if that state variable was not included in the initial `state` object.) This function will merge this object with the existing state object, overriding any existing values on the same variables. `setState` function calls are very often triggered by events that our users can trigger.
+**Why?**
+
+React depends on the hook functions being called in a specific order when a functional component is run.  It will not allow you to put code into your component which could affect the order in which `useState` is called.  You cannot even put `useState` inside an if statement.
+
+### 2.  You can only call hooks in functional components or other hooks
+
+Creating your own hooks is a more advanced topic, but state is intended to provide a way for a component to have information it remembers and changes internally over time.  So it does not make sense for state to be used outside of a component.  Therefore only call `useState` inside a React component.  Later if you create your own hooks, you can call them there.
```
Comment (jmaddox19) [inline React/state.md]:
```suggestion
State is intended to provide a way for a component to have information it remembers and changes internally over time, so it does not make sense for state to be used outside of a component.  Therefore only call `useState` inside a React component.  Later if you create your own hooks, you can call them there. But that's a topic we don't need to know about right now.
```
I found it confusing to start by addressing creating your own hooks, so I wanted to suggest removing that bit from the beginning of this line.

File: React/state.md [hunk 10]
```diff
@@ -1,93 +1,186 @@
...
+        />
+// ...
+```
+
+Then in our `Student` component we can change `useState(false)` and replace it with:
+
+```javascript
+// src/components/Student.js
+// ...
+  const [present, setPresent] = useState(props.present);
```
Comment (jmaddox19) [inline React/state.md]:
```suggestion
  const [isPresent, setIsPresent] = useState(props.present);
```
Only need this change if we change the one above.

File: React/class-components.md (hunk 1 above)
Comment (CheezItMan) [reply React/class-components.md]: We can definitely make this section optional.  I'll bring this up at our next standup.

💯

File: React/state.md (hunk 7 above)
Comment (CheezItMan) [reply React/state.md]: I was just thinking the student being present or absent, rather than making it two words.

File: React/events.md [hunk 11]
```diff
@@ -8,84 +8,112 @@
...
-    return (
-      ...
-      <button onClick={ this.onButtonClick }>Toggle Display</button>
-    );
-  }
+// src/components/Student.js
+// ...
+
+// Function to toggle present
+const onButtonClick = () => onButtonClick(!present);
```
Comment (jmaddox19) [inline React/events.md]: Should this be calling `setPresent`?
```suggestion
const onButtonClick = () => setPresent(!present);
```

File: React/events.md [hunk 12]
```diff
@@ -8,84 +8,112 @@
...
+
+  // Component functions always return JSX
+  return (
```
Comment (jmaddox19) [inline React/events.md]: I think it'd be helpful if we included the component definition for this code snippet so the context of the return is clear.
It took me a while to understand where this code was meant to fit in. (Even though in retrospect I do see you have a comment at the top to clarify that.

File: React/events.md [hunk 13]
```diff
@@ -8,84 +8,112 @@
...
+};
 ```
 
 Pay attention to the way we've named this callback. Starting event handler names with `on` is a standard convention, and lets other programmers know your intent. The name also gives information about:
+
 - What event happened (`change`)
 - Which DOM element was the target of the event (the name input)
 
-Next, we'll need to add the event handler function that our code will call when the button is clicked. Again, make note of the arrow function syntax.
+Next, we'll need to add the event handler function that our code will be called when the input is edited.
```
Comment (jmaddox19) [inline React/events.md]:
```suggestion
Next, we'll need to add a call to the event handler we just wrote.
```

File: React/events.md [hunk 14]
```diff
@@ -8,84 +8,112 @@
 React supplies support on many different events. We'll focus on a few of the common events, keyboard and click, but just know that you should reference the [full list](https://reactjs.org/docs/events.html#supported-events) if you want more information.
```
Comment (jmaddox19) [inline React/events.md]: Do you think it'd be worth adding screenshots to show how the website is changing as we make these changes? 
Our lessons aren't consistent about whether or not we do that, so I definitely don't think it's essential. It just occurred to me it's hard to follow what's happening in this lesson without seeing the visual changes. But maybe the expectation is that the student/teacher is following along with their own react project.

File: React/nested-event-handling.md [hunk 15]
```diff
@@ -14,156 +14,173 @@ By the end of this lesson, students should be able to...
...
-
-![nested components](images/nested-components.png)
-<!-- https://drive.google.com/open?id=1xq5jaCrI7FGp6PG1gr-bYE1ZTvPb5PxZ -->
-
-Note that we're passing in a new `isPresent` prop here. How could we modify our `Student` component to reflect this information?
+<details>
+  <summary>Our answer</summary>
 
-#### Removing State From `<Student />`
+  `setStudentList` a method `useState` provides to change the state first checks to see if the new state is different from the old state.  However it does not do a deep comparison for performance reasons.  Instead it checks to see if the new value references the same memory address as the old.  If so it does... NOTHING.  Therefore you need to pass in a new object to update state.  If you do not... nothing will change.
```
Comment (jmaddox19) [inline React/nested-event-handling.md]: Sorry, I'm having trouble understanding what this is saying

File: React/nested-event-handling.md [hunk 16]
```diff
@@ -14,156 +14,173 @@ By the end of this lesson, students should be able to...
...
 
 There are a few things to consider when making this change to our application. Think about these questions with your neighbors.
+
 1. Which component should contain the button that will enable us to mark an individual student as "present"?
 2. Which component is managing the student data?
 3. Is the answer to #1 the same as #2?
 
-If you answered NO to question #3, you're on to something big. The data related to the students is tracked in the `StudentCollection` while the button to mark an individual student "present" should really be on each individual `Student` component. Our challenge then is to use the tools we have been given to _propagate_ the button press event from one component to another.
+If you answered NO to question #3, you're on to something big. The data related to the students is tracked in `App` while the button to mark an individual student "present" should really be on each individual `Student` component. Our challenge then is to use the tools we have been given to _propagate_ the button press event from one component to another.
```
Comment (jmaddox19) [inline React/nested-event-handling.md]: I'm inclined to suggest hiding this with a details/summary block since it feels a lot like an answer to a question we want them to contemplate on their own

File: React/nested-event-handling.md [hunk 17]
```diff
@@ -14,156 +14,173 @@ By the end of this lesson, students should be able to...
...
 
 #### The CSS
```
Comment (jmaddox19) [inline React/nested-event-handling.md]: This bit feels to me like a confusing interjection. Would it be alright if we moved it down, after we've fully handled state and callbacks.

File: React/nested-event-handling.md [hunk 18]
```diff
@@ -180,187 +197,103 @@ The CSS is the most straightforward thing on our todo list, so let's get that ou
...
-```js
-// Student.js
-const Student = (props) => {
-
-return (
-  <section className='student'>
-    <h3>Student Component</h3>
-    <h4 className={(props.isPresent) ? 'present' : 'absent'}>Name {props.fullName} </h4>
-    <p>Email: {props.email} </p>
+#### Adding className for styling
```
Comment (jmaddox19) [inline React/nested-event-handling.md]: This would need to be moved along with it.

File: React/nested-event-handling.md [hunk 19]
```diff
@@ -14,156 +14,173 @@ By the end of this lesson, students should be able to...
...
+      <StudentCollection students={studentList} />
+    </div>
+  );
 }
 ```
 
-Next, we'll use a `map` in our `render` function to iterate through each item in our state object and render a `Student` component for each piece of data. You should already have some of this `map` logic, but now we'll be getting the data from `state` rather than a constant.
+We can now verify that the app still works as normal.  
+
+We can also add a function to change a student.  To make things easier we can add an `id` field to each student object.
```
Comment (jmaddox19) [inline React/nested-event-handling.md]:
```suggestion
We can also add a function to change a student.  In addition, to make things easier we can add an `id` field to each student object.
```

File: React/nested-event-handling.md [hunk 20]
```diff
@@ -180,187 +197,103 @@ The CSS is the most straightforward thing on our todo list, so let's get that ou
...
+      fullName: props.fullName,
+      birthday: props.birthday,
+      email: props.email,
+      // Toggle present to a new value
+      present: !props.present,
+      id: props.id,
+    }
+
+    // call the function passed from `App`
+    props.onUpdateStudent(updatedStudent);
```
Comment (jmaddox19) [inline React/nested-event-handling.md]: This callback right here feels like a monumental new concept to students. Can we give more explanation to explicitly state this is a call to the function we created in `App`?

File: React/nested-event-handling.md [hunk 21]
```diff
@@ -180,187 +197,103 @@ The CSS is the most straightforward thing on our todo list, so let's get that ou
...
-  disabled={ props.isPresent }
-  onClick={ onPresentButtonClick }
-  >Mark Present </button>
-  );
-}
-```
-
-Test it out and see what gets logged in the console.
-
 **Question:** What piece of data is going to help us determine exactly which student should be updated?
```
Comment (jmaddox19) [inline React/nested-event-handling.md]:
```suggestion
**Question:** Once we call `onUpdateStudent`, execution will be passed back to the `App` component. What piece of data is going to help the `App` component determine exactly which student should be updated?
```

File: React/nested-event-handling.md [hunk 22]
```diff
@@ -180,187 +197,103 @@ The CSS is the most straightforward thing on our todo list, so let's get that ou
...
+`onButtonClick` is closed around `props`, so the `props` variable is still accessible! Remember, all we need to close is three things:
 
 1. Nest a function inside a function
 1. Reference a variable from the outer function in the inner function
 1. Make the inner function available outside the outer function
 
-1. `onPresentButtonClick` is nested inside of `Student`
-1. We reference `props` inside of `onPresentButtonClick`
-1. We hand `onPresentButtonClick` to the `<button>`, where it lives on after `Student` has finished running!
+1. `onButtonClick` is nested inside of `Student`
```
Comment (jmaddox19) [inline React/nested-event-handling.md]: I think this is intended to be 1,2,3 but it's rendering as 4,5,6. Perhaps an additional newline would fix this problem.

File: React/nested-event-handling.md [hunk 23]
```diff
@@ -180,187 +197,103 @@ The CSS is the most straightforward thing on our todo list, so let's get that ou
...
 
 ![State management with callbacks](images/state-management-with-callbacks.png)
 
-<!-- https://drive.google.com/open?id=1byKvDyUP5HUwQojmg2cpL4eNO9keFYpD -->
+<!-- https://drive.google.com/file/d/1dyNbUUe_6rkrSEG_lPKJGiwAAOGIHQ5_/view?usp=sharing -->
 
 **On the left in orange** is the setup. Information is passed from parent components to child components at `render` time using `props`.
+
+- `App` provides a list of students to `StudentCollection`
 - `StudentCollection` gives `Student`:
```
Comment (jmaddox19) [inline React/nested-event-handling.md]: We need to adjust this since there isn't a `StudentCollection` anymore.

**Comment by jmaddox19** [comment]
Just skimmed state-with-classes and have no comments to add :)

File: React/events.md (hunk 12 above)
Comment (CheezItMan) [reply React/events.md]: done

File: React/events.md (hunk 14 above)
Comment (CheezItMan) [reply React/events.md]: I think the expectation is that the teacher is doing this along with the student, and it's hard to see the result because this lesson is based on events, but I can add a pair of'em.

Comment (CheezItMan) [reply React/events.md]: And done

File: React/nested-event-handling.md (hunk 23 above)
Comment (CheezItMan) [reply React/nested-event-handling.md]: Actually there is it forwards individual student data to each student component.

File: React/nested-event-handling.md (hunk 22 above)
Comment (CheezItMan) [reply React/nested-event-handling.md]: Fixed, and thanks!

File: React/nested-event-handling.md (hunk 20 above)
Comment (CheezItMan) [reply React/nested-event-handling.md]: There's quite a bit on this below, but I've added more.

**Comment by jmaddox19** [comment]
@CheezItMan Thank you for responding to so much!
There are 4 comments Github decided to collapse for some reason. It looks like you might not have seen those

File: React/nested-event-handling.md (hunk 15 above)
Comment (jmaddox19) [reply React/nested-event-handling.md]: I think you missed this since github decided to collapse it.

File: React/nested-event-handling.md (hunk 16 above)
Comment (jmaddox19) [reply React/nested-event-handling.md]: I think you missed this since github decided to collapse it.

File: React/nested-event-handling.md (hunk 17 above)
Comment (jmaddox19) [reply React/nested-event-handling.md]: I think you missed this since github decided to collapse it.

File: React/nested-event-handling.md (hunk 18 above)
Comment (jmaddox19) [reply React/nested-event-handling.md]: I think you missed this since github decided to collapse it.

File: React/class-components.md (hunk 1 above)
Comment (kaidamasaki) [reply React/class-components.md]: I really like making this an optional reading.

File: React/events.md [hunk 24]
```diff
@@ -8,84 +8,119 @@
...
+        </button>
+      </div>
     );
-  }
+}
 ```
 
+**Pause here** for a moment with the person sitting next to you. What is this code doing? Refer back to the [state lecture](state.md) if you can't remember what the `setPresent` function is supposed to do.
+
+**Note** the arrow function syntax for this event handler function. You could define the event handler function `onButtonClick` as a regular function, but making it an arrow function in this case is a bit more condense and keeps the `this` context to the current component.
```
Comment (kaidamasaki) [inline React/events.md]:
```suggestion
**Note** the arrow function syntax for this event handler function. You could define the event handler function `onButtonClick` as a regular function, but making it an arrow function in this case is a bit more concise and keeps the `this` context to the current component.  
```

Comment (kaidamasaki) [reply React/events.md]: Note: `this` is far less meaningful when using hooks instead of classes and may not be worth mentioning here.

File: React/events.md [hunk 25]
```diff
@@ -8,84 +8,119 @@
...
-Within our component (usually after the `constructor` and before the `render`), we'll add the event handler function that our code will call when the button is clicked.  
+Another variable that we store is the name. Unfortunately right now, we are passing the `fullName` from props and we **cannot** change the value of our props.  
 
-```javascript
-// between constructor and render
-onButtonClick = () => {
-  this.setState({ displayName: !this.state.displayName });
-}
-```
+**1.  Convert `fullName` from using props to state.**
```
Comment (kaidamasaki) [inline React/events.md]: Maybe make this an H5?

```suggestion
##### 1.  Convert `fullName` from using props to state.
```

File: React/nested-event-handling.md (hunk 15 above)
Comment (CheezItMan) [reply React/nested-event-handling.md:124]: Ok I've rephrased it.

**Comment by CheezItMan** [comment]
@jmaddox19 Thanks Jared I've responded to those hidden comments and I think it flows much better now.
